skafka/src/test/scala/com/evolutiongaming/skafka/consumer/ListenersLockTest.scala

```diff
@@ -0,0 +1,283 @@
+import org.apache.kafka.common.{Metric, MetricName, PartitionInfo, TopicPartition => TopicPartitionJ}
+import org.scalatest.funsuite.AsyncFunSuite
+import org.scalatest.matchers.should.Matchers
+
+import scala.concurrent.duration._
+import scala.jdk.CollectionConverters._
+
+class ListenersLockTest extends AsyncFunSuite with Matchers {
+
+  test("listenersLock") {
+    listenersLock[IO].run()
+  }
+
+  test("consumer.poll") {
+    `consumer.poll`[IO].run()
+  }
+
+  private def listenersLock[F[_]: Concurrent] = {
+    ().pure[F] // TODO
+  }
+
+  private def `consumer.poll`[F[_]: Concurrent: ToTry: ToFuture: Blocking: ContextShift] = {
+
+    val topic = "topic"
+
+    sealed trait Action
+
+    object Action {
+      case object PartitionsAssignedEnter extends Action
+      case object PartitionsAssignedExit extends Action
+      case object PartitionsRevokedEnter extends Action
+      case object PartitionsRevokedExit extends Action
+      case object Topics extends Action
+    }
+
+    trait Actions {
+
+      def add(action: Action): F[Unit]
+
+      def get: F[List[Action]]
+    }
+
+    object Actions {
+      def of: F[Actions] = {
+        Ref[F]
+          .of(List.empty[Action])
+          .map { ref =>
+            new Actions {
+              def add(action: Action) = ref.update { action :: _ }
+
+              def get = ref.get.map { _.reverse }
+            }
+          }
+      }
+    }
+
+    def consumerJ[K, V](actions: Actions): F[ConsumerJ[K, V]] = {
+      for {
+        listenerRef <- Ref[F].of(none[ConsumerRebalanceListener])
+      } yield {
+        val consumer: ConsumerJ[K, V] = new ConsumerJ[K, V] {
+
+          def assignment() = Set.empty[TopicPartitionJ].asJava
+
+          def subscription() = Set.empty[String].asJava
+
+          def subscribe(topics: CollectionJ[String]) = {}
+
+          def subscribe(topics: CollectionJ[String], listener: ConsumerRebalanceListener) = {
+            listenerRef
+              .set(listener.some)
+              .toTry
+              .get
+          }
+
+          def assign(partitions: CollectionJ[TopicPartitionJ]) = {}
+
+          def subscribe(pattern: Pattern, callback: ConsumerRebalanceListener) = {}
+
+          def subscribe(pattern: Pattern) = {}
+
+          def unsubscribe() = {}
+
+          def poll(timeout: Long) = poll(Duration.ofMillis(timeout))
+
+          def poll(timeout: Duration) = {
+            listenerRef
+              .get
+              .flatMap { listener =>
+                listener.foldMapM { listener =>
+                  def partitions(value: Int) = {
+                    List(TopicPartition(topic, Partition.unsafe(value)))
+                      .map { _.asJava }
+                      .asJava
+                  }
+                  for {
+                    _ <- Sync[F].delay { listener.onPartitionsAssigned(partitions(0)) }
```

Comment:
Let's check in the test that the reverse (e.g. default) order of assigned/revoke also holds.

Reply:
no need anymore :)